4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py

```diff
@@ -0,0 +1,283 @@
+                    result_row = {
+                        "z": int(col),
+                        "full_model_z": num_components,
+                        "model": str(model_name),
+                        "reactome pathway": str(pathway_result['Term']),
+                        "gsea es score": pathway_result['es'],
+                        "nes score": pathway_result['nes'],
+                        "p value": pathway_result['pval'],
+                        "shuffled": False
+                    }
+                    gsea_results.append(result_row)
+                    
+            except ZeroDivisionError:
+                print(f"Skipping GSEA for {col} due to zero division error.")
+    
+    # Convert GSEA results into a DataFrame and save to a Parquet file
+    gsea_results_df = pd.DataFrame(gsea_results)
+    # Filter for significant results
+    significant_gsea_df = gsea_results_df[
+            (gsea_results_df['gsea es score'].abs() > lfc_cutoff) &
+            (gsea_results_df['p value'] < fdr_cutoff)
+        ]
+    print(significant_gsea_df)
+    
+    return significant_gsea_df
+
+
+# Initialize a list to store the final results
+final_gsea_results = []
+
+# GSEA settings
+latent_dims = [2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 18, 20, 25, 30, 35, 40, 45, 50, 60, 70, 80, 90, 100, 150]
+model_names = ["pca", "ica", "nmf", "vanillavae", "betavae", "betatcvae"]
+
+# Define the output file path
+final_output_file = pathlib.Path(output_dir) / "combined_z_matrix_gsea_results.parquet"
+
+# Try to load the existing combined results DataFrame if it exists
+try:
+    combined_results_df = pd.read_parquet(final_output_file)
+    print(f"Loaded existing combined results from {final_output_file}")
+except FileNotFoundError:
+    # If the file doesn't exist, initialize an empty DataFrame
+    combined_results_df = pd.DataFrame()
+    print(f"No existing file found. Initialized empty DataFrame.")
+
+for num_components in latent_dims:
+    for model_name in model_names:
+        # Check if this model and latent dimension have already been processed
+        if not combined_results_df.empty:
+            if ((combined_results_df['model'] == model_name) & 
+                (combined_results_df['full_model_z'] == num_components)).any():
+                print(f"Skipping {model_name} with {num_components} dimensions as it is already processed.")
+                continue  # Skip to the next iteration if this combination is already present
+        z_matrix_df = None
+        
+        if model_name in ["pca", "ica", "nmf"]:
+            # Sklearn models (PCA, ICA, NMF)
+            if model_name == "pca":
+                model = PCA(n_components=num_components)
+            elif model_name == "ica":
+                model = FastICA(n_components=num_components)
+            elif model_name == "nmf":
+                model = NMF(n_components=num_components, init='nndsvd', max_iter=1000, random_state=0)
```

Comment:
Any specific reason to use this `init` method? (just curious!)

Reply:
At a certain number of components the NMF model was running into sparseness errors, which is best for the nndsvd initialization rather than random according to the documentation